File: backend/src/main/java/edu/university/grademanagement/model/entity/Course.java
package edu.university.grademanagement.model.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

/**
 * Course Entity
 * Represents a course in the university
 */
@Entity
@Table(name = "COURSES", schema = "GMS_ADMIN")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Course {

    @Id
    @Column(name = "course_id", length = 10)
    private String courseId;

    @Column(name = "course_name", nullable = false, length = 100)
    private String courseName;

    @Column(name = "credits", nullable = false)
    private Integer credits;

    @Column(name = "department_id", nullable = false, length = 10)
    private String departmentId;

    @Column(name = "course_type", length = 20)
    private String courseType;

    @Column(name = "prerequisite_course_id", length = 10)
    private String prerequisiteCourseId;

    @Column(name = "description", length = 500)
    private String description;

    @Column(name = "created_date")
    private LocalDateTime createdDate;

    @Column(name = "updated_date")
    private LocalDateTime updatedDate;

    @PrePersist
    protected void onCreate() {
        createdDate = LocalDateTime.now();
        updatedDate = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedDate = LocalDateTime.now();
    }

    public boolean isMandatory() {
        return "Mandatory".equals(courseType);
    }

    public boolean isElective() {
        return "Elective".equals(courseType);
    }

    public boolean hasPrerequisite() {
        return prerequisiteCourseId != null && !prerequisiteCourseId.isEmpty();
    }
}

File: flutter_app/lib/models/course.dart
class Course {
  final String? courseId;
  final String? courseName;
  final int? credits;
  final String? departmentId;
  final String? courseType;
  final String? prerequisiteCourseId;
  final String? description;
  final DateTime? createdDate;
  final DateTime? updatedDate;

  Course({
    this.courseId,
    this.courseName,
    this.credits,
    this.departmentId,
    this.courseType,
    this.prerequisiteCourseId,
    this.description,
    this.createdDate,
    this.updatedDate,
  });

  factory Course.fromJson(Map<String, dynamic> json) {
    return Course(
      courseId: json['courseId'] as String?,
      courseName: json['courseName'] as String?,
      credits: json['credits'] as int?,
      departmentId: json['departmentId'] as String?,
      courseType: json['courseType'] as String?,
      prerequisiteCourseId: json['prerequisiteCourseId'] as String?,
      description: json['description'] as String?,
      createdDate: json['createdDate'] != null
          ? DateTime.tryParse(json['createdDate'])
          : null,
      updatedDate: json['updatedDate'] != null
          ? DateTime.tryParse(json['updatedDate'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'courseId': courseId,
      'courseName': courseName,
      'credits': credits,
      'departmentId': departmentId,
      'courseType': courseType,
      'prerequisiteCourseId': prerequisiteCourseId,
      'description': description,
    };
  }

  bool get isMandatory => courseType == 'Mandatory';
  bool get isElective => courseType == 'Elective';
  bool get hasPrerequisite => prerequisiteCourseId != null && prerequisiteCourseId!.isNotEmpty;

  Course copyWith({
    String? courseId,
    String? courseName,
    int? credits,
    String? departmentId,
    String? courseType,
    String? prerequisiteCourseId,
    String? description,
  }) {
    return Course(
      courseId: courseId ?? this.courseId,
      courseName: courseName ?? this.courseName,
      credits: credits ?? this.credits,
      departmentId: departmentId ?? this.departmentId,
      courseType: courseType ?? this.courseType,
      prerequisiteCourseId: prerequisiteCourseId ?? this.prerequisiteCourseId,
      description: description ?? this.description,
    );
  }
}

